library/nist_kw.c

```diff
@@ -0,0 +1,710 @@
+                                                 MBEDTLS_DECRYPT )
+                                                                   ) != 0 )
+    {
+        return( ret );
+    }
+
+    return( 0 );
+}
+
+/*
+ * Free context
+ */
+void mbedtls_nist_kw_free( mbedtls_nist_kw_context *ctx )
+{
+    mbedtls_cipher_free( &ctx->cipher_ctx );
+    mbedtls_platform_zeroize( ctx, sizeof( mbedtls_nist_kw_context ) );
+}
+
+/*
+ * Helper function for Xoring the uint64_t "t" with the encrypted A.
+ * Defined in NIST SP 800-38F section 6.1
+ */
+static void calc_a_xor_t( unsigned char A[KW_SEMIBLOCK_LENGTH], uint64_t t )
+{
+    size_t i = 0;
+    for( i = 0; i < sizeof( t ); i++ )
+    {
+        A[i] ^= ( t >> ( ( sizeof( t ) - 1 - i ) * 8 ) ) & 0xff;
+    }
+}
+
+/*
+ * KW-AE as defined in SP 800-38F section 6.2
+ * KWP-AE as defined in SP 800-38F section 6.3
+ */
+int mbedtls_nist_kw_wrap( mbedtls_nist_kw_context *ctx,
+                          mbedtls_nist_kw_mode_t mode,
+                          const unsigned char *input, size_t in_len,
+                          unsigned char *output, size_t *out_len )
+{
+    int ret = 0;
+    size_t semiblocks = 0;
+    size_t s;
+    size_t olen, padlen = 0;
+    uint64_t t = 0;
+    unsigned char outbuff[KW_SEMIBLOCK_LENGTH * 2];
+    unsigned char inbuff[KW_SEMIBLOCK_LENGTH * 2];
+    unsigned char *R2 = output + KW_SEMIBLOCK_LENGTH;
+    unsigned char *A = output;
+    const uint32_t ICV1[] = NIST_KW_ICV1;
+    const uint32_t ICV2[] = NIST_KW_ICV2;
+    *out_len = 0;
+    /*
+     * Generate the String to work on
+     */
+    if( mode == MBEDTLS_KW_MODE_KW )
+    {
+        /*
+         * According to SP800-38F Table 1, the plaintext length for KW
+         * must be between 2 to 2^54-1 semiblocks inclusive.
+         */
+        if( in_len < 8
+#if SIZE_MAX > 0x1FFFFFFFFFFFFF8
+            || in_len > 0x1FFFFFFFFFFFFF8
+#endif
+           )
+        {
+            ret = MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+            goto cleanup;
+        }
+
+
+        memcpy( output, ICV1, KW_SEMIBLOCK_LENGTH );
+        memcpy( output + KW_SEMIBLOCK_LENGTH, input, in_len );
+    }
+    else
+    {
+        /*
+         * According to SP 800-38F Table 1, the plaintext length for KWP
+         * must be between 1 and 2^32-1 octets inclusive.
+         */
+        if( in_len < 1
+#if SIZE_MAX > 0xFFFFFFFF
+            || in_len > 0xFFFFFFFF
+#endif
+          )
+        {
+            ret = MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+            goto cleanup;
+        }
+
+        if( in_len % 8 != 0 )
+        {
+            padlen = ( 8 - ( in_len % 8 ) );
+        }
+
+        memcpy( output, &ICV2, KW_SEMIBLOCK_LENGTH / 2 );
+        PUT_UINT32_BE( ( in_len & 0xffffffff ), output,
+                       KW_SEMIBLOCK_LENGTH / 2 );
+
+        memcpy( output + KW_SEMIBLOCK_LENGTH, input, in_len );
+        memset( output + KW_SEMIBLOCK_LENGTH + in_len, 0, padlen );
```

Comment:
This is documented, and regarded as user error, if buffer not big enough

Reply:
That's not very good. It's better to have the checking in code instead of only in documentation especially as the requirements on the length are not trivial and vary based on KW or KWP modes; that's a lot to ask of the user and will lead to more mistakes than necessary. In most places in Mbed TLS where there is no separate output buffer size, the output buffer is the same size as the input buffer and so the provided length works for both in and out buffers. (`mbedtls_hkdf()` is one recent example where we have size fields for all supplied buffers, whether input or output.)

Reply:
For example, in `pk_sign()`, we get an output parameter as `sig_len` and the signature buffer isn't being checked.
I agree that we should do parameter checking, but you know the approach on that matter.

Reply:
We recognize that the signature of `pk_sign` is not good and are planning to change it in the next major version when we break source code compatibility.

Reply:
ok, so `mbedtls_cipher_update()` is another example.

However, I think it's worth having checks against user mistakes, but I think we should use `out_len` as both an in \ out parameter, to indicate buffer size as input, and bytes written as output, as this is much more user friendly and readable than having another parameter for output buffer size